### ISSUE TITLE
Changes to cart item/WC Product stock aren't used for variations in cart/checkout blocks (but are for shortcode cart/checkout)

### ISSUE THREAD

**Issue by ninetyninew** [woothemes/woocommerce#44770]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

If you make changes to cart item data, such as stock, then these changes aren't considered in cart/checkout for variations in the cart, specifically on cart/checkout blocks (this is different to shortcode cart/checkout which does). Also note that it does work fine for simple products on cart/checkout blocks, just not variations.

To show this occurring we have created a code snippet, this changes stock data on a cart item (and also does it on cart/checkout for WC_Product and WC_Product_Variation based returns too, just to check that isn't the issue).

You'll find when following the reproduce steps below that variations will not allow checkout as throws an error that there isn't enough stock, yet we are specifically telling it that the product is in stock in the code snippet. 

### Expected behavior

Cart/checkout blocks take cart item data changes for stock into account for variation products, the same as simple products do.

### Actual behavior

Cart/checkout blocks does not take cart item data changes for stock into account for variation products (works fine for simple products). This is different to shortcode cart/checkout blocks which do.

### Steps to reproduce

- Add this code snippet to functions.php
- Read the comments to understand what it is doing (some of this is just for the purpose of highlighting the issue)

```
function cart_item_changes() {

	// This changes the cart item product objects to allow checkout of higher stock levels than the standard stock qty field

	$cart = WC()->cart;

	if ( !empty( $cart ) ) {

		if ( isset( $cart->cart_contents ) ) {

			if ( !empty( $cart->cart_contents ) ) {

				foreach ( $cart->cart_contents as $cart_content_key => $cart_content_value ) {

					// We'd normally have some condition here to determine if the stock data changes should be used

					$cart_content_value['data']->set_manage_stock( false );
					$cart_content_value['data']->set_stock_status( 'instock' );
					$cart_content_value['data']->set_stock_quantity( PHP_INT_MAX );

					// Above are set to bypass the Woo stock validation, we then deal with determining if available to order later in the process through alternative means, but for the purpose of this example the above should let the order be placed if product has 1 stock, and the cart qty changed to 10 (aware the customer can only select 1 on the product page, but that is amended via other functionality to allow up to max of other stock data depending on if they enabled a checkbox on the product page which means the stock changes used instead)

				}

			}

		}

	}

}
add_action( 'woocommerce_cart_updated', 'cart_item_changes', PHP_INT_MAX );
add_action( 'woocommerce_before_calculate_totals', 'cart_item_changes', PHP_INT_MAX );


function product_return_changes( $value, $product ) {

	// This is also included as we found shortcode cart/checkout needs it, which effectively filters the product object returns if got via the cart/checkout, assumedly as some data in that scenario is got from the non-cart item product objects

	if ( is_cart() || is_checkout() ) {

		// We'd normally have some condition here to determine whether the stock changes we are doing should be used, and if so do the following:

		$is_in_stock = true;
		$stock_quantity = PHP_INT_MAX;
		$manage_stock = false;

		// Above are set for the same bypass reasons as in cart_item_changes()

	}

	if ( ( 'woocommerce_product_is_in_stock' == current_filter() || 'woocommerce_product_variation_is_in_stock' == current_filter() ) && isset( $is_in_stock ) ) {

		return $is_in_stock;

	} elseif ( ( 'woocommerce_product_get_stock_quantity' == current_filter() || 'woocommerce_product_variation_get_stock_quantity' == current_filter() ) && isset( $stock_quantity ) ) {

		return $stock_quantity;

	} elseif ( ( 'woocommerce_product_get_manage_stock' == current_filter() || 'woocommerce_product_variation_get_manage_stock' == current_filter() ) && isset( $manage_stock ) ) {

		return $manage_stock;

	} else {

		return $value;

	}

}

add_filter( 'woocommerce_product_is_in_stock', 'product_return_changes', PHP_INT_MAX, 2 );
add_filter( 'woocommerce_product_variation_is_in_stock', 'product_return_changes', PHP_INT_MAX, 2 );
add_filter( 'woocommerce_product_get_stock_quantity', 'product_return_changes', PHP_INT_MAX, 2 );
add_filter( 'woocommerce_product_variation_get_stock_quantity', 'product_return_changes', PHP_INT_MAX, 2 );
add_filter( 'woocommerce_product_get_manage_stock', 'product_return_changes', PHP_INT_MAX, 2 );
add_filter( 'woocommerce_product_variation_get_manage_stock', 'product_return_changes', PHP_INT_MAX, 2 );

function hold_stock( $minutes ) {

	// We'd normally have some condition here to determine whether items in cart should use our changed stock data, and if so make this 0, to eliminate "Not enough units of xxx are available in stock to fulfil this order."

	return 0;

}
add_filter( 'option_woocommerce_hold_stock_minutes', 'hold_stock' );

function allow_cart_qty_change( $value, $product, $cart_item ) {

	// This is just here for this example to allow you to select more than the stock in cart to test placing an order with a qty higher than the standard stock, normally the quantity here would have been passed from the product page based on if the stock data changes we are doing are required

	return PHP_INT_MAX;

}
add_filter( 'woocommerce_store_api_product_quantity_maximum', 'allow_cart_qty_change', PHP_INT_MAX, 3 );
```

- Create simple product, manage stock on, no backorders, stock 1
- Create variable product, manage stock on, no backorders, stock 1

- Make sure cart empty
- Add the simple product to cart
- Up the quantity to more than 1, see `allow_cart_qty_change()` in our snippet why
- Place the order, goes through fine
- Ensure cart empty (should be as order just placed)
- Add the variation product to cart
- Up the quantity to more than 1, `see allow_cart_qty_change()` in our snippet why
- Place the order
- See error `There is not enough "xxx" in stock. Please reduce the quantity in your cart.`

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://luckyshow.s4-tastewp.com
Site address (URL): https://luckyshow.s4-tastewp.com
WC Version: 8.6.0
REST API Version: ✔ 8.6.0
WC Blocks Version: ✔ 11.8.0-dev
Action Scheduler Version: ✔ 3.7.1
Log Directory Writable: ✔
WP Version: 6.4.3
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: TasteWP-S4 Official/3.0.0
PHP Version: 8.0.30
PHP Post Max Size: 128 MB
PHP Time Limit: 100
PHP Max Input Vars: 5000
cURL Version: 7.68.0
OpenSSL/1.1.1f

SUHOSIN Installed: –
MySQL Version: 8.0.32-0ubuntu0.20.04.2
Max Upload Size: 60 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 8.6.0
WC Database Prefix: wp_
Total Database Size: 3.21MB
Database Data Size: 1.33MB
Database Index Size: 1.88MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.05MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_litespeed_url: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_litespeed_url_file: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_options: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.13MB + Index: 0.11MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_snippets: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.05MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_addresses: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_operational_data: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_orders: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_wc_orders_meta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wcrp_rental_products_rentals: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wcrp_rental_products_rentals_archive: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wpfm_backup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB

### Post Type Counts ###

attachment: 24
page: 8
post: 3
product: 21
product_variation: 12
revision: 1
shop_order_placehold: 2

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (2) ###

Code Snippets: by Code Snippets Pro – 3.6.2
WooCommerce: by Automattic – 8.6.0

### Dropin Plugins (1) ###

maintenance.php: maintenance.php

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to Woo.com: –
Enforce Approved Product Download Directories: ✔
HPOS feature screen enabled: ✔
HPOS feature enabled: ✔
Order datastore: Automattic\WooCommerce\Internal\DataStores\Orders\OrdersTableDataStore
HPOS data sync enabled: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: GeneratePress
Version: 3.4.0
Author URL: https://tomusborne.com
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
product-variation-management
product-virtual-downloadable
product-external-affiliate
product-grouped
product-linked
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: customize-store
minified-js
new-product-management-experience
product-pre-publish-modal
settings
async-product-editor-category-field

Daily Cron: ✔ Next scheduled: 2024-02-20 10:16:43 +00:00
Options: ✔
Notes: 65
Onboarding: skipped

### Action Scheduler ###

Complete: 52
Oldest: 2024-02-19 10:17:47 +0000
Newest: 2024-02-19 12:07:37 +0000

Failed: 1
Oldest: 2024-02-19 10:17:47 +0000
Newest: 2024-02-19 10:17:47 +0000

Pending: 1
Oldest: 2024-02-20 10:17:47 +0000
Newest: 2024-02-20 10:17:47 +0000

### Status report information ###

Generated at: 2024-02-19 12:26:00 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by opr** [comment]
In [Slack](https://woocommercecommunity.slack.com/archives/C03MCL9RTAR/p1709291644483929?thread_ts=1709218584.859379&cid=C03MCL9RTAR) I spoke with @ninetyninew and we have agreed to leave this as `needs: author feedback` until further debugging has been carried out.

@ninetyninew when you have more info you can reply here and we will proceed to triage this. Thanks!

**Comment by ninetyninew** [comment]
@opr Upon further investigation and further understanding of the changes which have been made to the reserve stock mechanism specifically for cart/checkout blocks, I can confirm that this issue should be disregarded.

For clarity on what was found for others who may experience this, I found that for variations I was assuming that by overriding `manage_stock` on the variation cart item to `false` then manage stock would be off, but actually this being false during Woo's cart validation checks make this actually return `parent`, and then it goes and gets the `manage_stock` of the parent, so it was validating based off the parent stock settings.

This was also combined with the issue of a new hardcoded 10 minute stock reservation for cart/checkout block entry, [described here](https://github.com/woocommerce/woocommerce/issues/45245) which I have also [raised a PR](https://github.com/woocommerce/woocommerce/pull/45246) regarding. As the stock reservation ultimately gets the overall product object (not a cart item product object), and from there discovers the product is out of stock, even though we are attempting to force it to be in stock. Before the hardcoded 10 minutes, we could disable the stock reservation, but for cart/checkout blocks this isn't possible currently due to the forced 10 minutes, hence why I have raised a PR to add a filter to the 10 minute duration.

This explains why it worked on shortcode cart/checkout but not blocks.
